src/components/controls/OpacityControl.js

```diff
@@ -0,0 +1,323 @@
+    this.maxLevel = 1;
+    this.opCanvas = null;
+    this.opContext = null;
+
+    this.dragging = false;
+    this.hovering = false;
+    this.nodeDragged = -1;
+    this.nodeHovered = -1;
+    this.dragStart = [0, 0];
+    this.startPos = [0, 0];
+    this.height = 70;
+    this.padding = 10;
+    // distance where clicks and hovering near control nodes are registered
+    this.hoverRadius = 15;
+    this.width = 0;
+
+    this.nodes = [
+      { x: 0, y: 0 },
+      { x: 250 * 0.11, y: 15 },
+      { x: 250 * 0.32, y: 35 },
+      { x: 250 * 0.92, y: 70 },
+    ];
+
+    this.nodesCanvasSpace = [];
+
+    this.changePointer = this.changePointer.bind(this);
+    this.onMouseDown = this.onMouseDown.bind(this);
+    this.onMouseUp = this.onMouseUp.bind(this);
+    this.dragPointer = this.dragPointer.bind(this);
+    this.addPoint = this.addPoint.bind(this);
+    this.removePoint = this.removePoint.bind(this);
+    this.sendAlphaData = this.sendAlphaData.bind(this);
+    this.resetOpacityPoints = this.resetOpacityPoints.bind(this);
+  }
+
+  componentDidMount() {
+    this.updateCanvas();
+    document.addEventListener("mousemove", this.dragPointer);
+    document.addEventListener("mouseup", this.onMouseUp);
+
+    this.opCanvas.addEventListener("mousemove", this.changePointer);
+    this.opCanvas.addEventListener("mousedown", this.onMouseDown);
+    this.opCanvas.addEventListener("dblclick", this.addPoint);
+    this.opCanvas.addEventListener("contextmenu", this.removePoint);
+  }
+
+  componentWillUnmount() {
+    this.opCanvas.removeEventListener("mousemove", this.changePointer);
+    document.removeEventListener("mousemove", this.dragPointer);
+    this.opCanvas.removeEventListener("mousedown", this.onMouseDown);
+    document.removeEventListener("mouseup", this.onMouseUp);
+    this.opCanvas.removeEventListener("dblclick", this.addPoint);
+    this.opCanvas.removeEventListener("contextmenu", this.removePoint);
+    //-- Save state
+  }
+
+  updateCanvas() {
+    this.opCanvas = this.refs.canvas;
+    this.opContext = this.refs.canvas.getContext("2d");
+
+    const controlsContainer = document.getElementById("controls");
+
+    this.opCanvas.width = controlsContainer.clientWidth - 2 * this.padding - 2;
+    this.opCanvas.height = this.height + this.padding * 2;
+
+    this.opCanvas.width = 250;
+    this.width = this.opCanvas.width - 2 * this.padding;
+
+    this.opCanvas.style.border = "1px solid";
+    this.opContext.clearRect(0, 0, this.opCanvas.width, this.opCanvas.height);
+
+    this.opContext.strokeStyle = "rgba(128, 128, 128, 0.8)";
+    this.opContext.lineWidth = 2;
+
+    this.nodesCanvasSpace = [];
+
+    for (let i = 0; i < this.nodes.length; i++) {
+      let xPosInvertedCanvas = ~~this.nodes[i].x + this.padding;
+      let yPosInvertedCanvas = ~~(this.height - this.nodes[i].y) + this.padding;
+      this.nodesCanvasSpace.push({
+        x: xPosInvertedCanvas,
+        y: yPosInvertedCanvas,
+      });
+    }
```

Comment:
Maybe a [`for...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) loop here?

Reply:
Yeah that's what I figured. `OpacityControls.js` is going to be tackled in a separate PR